moveit_planners/ompl/ompl_interface/test/test_constrained_state_validity_checker.cpp

```diff
@@ -0,0 +1,326 @@
+    result_2 = checker->isValid(ompl_state.get(), distance_2);
+    EXPECT_FALSE(result_2);
+    // isValid function returned false before any distance calculation is done
+    EXPECT_EQ(distance_2, 0.0);
+  }
+
+  void testSelfCollision(const std::vector<double>& position_in_self_collision)
+  {
+    // create a validity checker for this test
+    auto checker = std::make_shared<ompl_interface::ConstrainedPlanningStateValidityChecker>(planning_context_.get());
+    checker->setVerbose(VERBOSE);
+
+    robot_state_->setJointGroupPositions(joint_model_group_, position_in_self_collision);
+
+    // use a scoped OMPL state so we don't have to call allocState and freeState
+    // (as recommended in the OMPL documantion)
+    ompl::base::ScopedState<> ompl_state(constrained_state_space_);
+    state_space_->copyToOMPLState(ompl_state.get(), *robot_state_);
+
+    // too much code, just to print the state
+    if (VERBOSE)
+    {
+      // cast ompl state to a specific type, useful in the rest of this test
+      auto state = ompl_state->as<ompl::base::ConstrainedStateSpace::StateType>()
+                       ->getState()
+                       ->as<ompl_interface::ConstrainedPlanningStateSpace::StateType>();
+
+      // ompl_state.reals() throws a segmentation fault for this state type
+      // use a more involved conversion to std::vector for logging
+      ROS_INFO_STREAM(std::vector<double>(state->values, state->values + joint_model_group_->getVariableCount()));
+    }
+
+    // the given state is known to be in self-collision, we check it here
+    bool result = checker->isValid(ompl_state.get());
+    EXPECT_FALSE(result);
+
+    // do the same for the version with distance
+    double distance{ 0.0 };
+    result = checker->isValid(ompl_state.get(), distance);
+    EXPECT_FALSE(result);
+
+    // but it should respect the joint limits
+    bool result_2 = robot_state_->satisfiesBounds();
+    EXPECT_TRUE(result_2);
+  }
+
+  // /***************************************************************************
+  //  * END Test implementation
+  //  * ************************************************************************/
+
+protected:
+  void SetUp() override
+  {
+    // setup all the input we need to create a StateValidityChecker
+    setupStateSpace();
+    setupPlanningContext();
+  };
+
+  void TearDown() override
+  {
+  }
+
+  void setupStateSpace()
+  {
+    ompl_interface::ModelBasedStateSpaceSpecification space_spec(robot_model_, group_name_);
+    state_space_ = std::make_shared<ompl_interface::ConstrainedPlanningStateSpace>(space_spec);
+    state_space_->computeLocations();  // this gets called in the state space factory normally
+
+    assert(state_space_ != nullptr);
+
+    auto dummy_constraint = std::make_shared<DummyConstraint>(num_dofs_);
+    constrained_state_space_ = std::make_shared<ompl::base::ProjectedStateSpace>(state_space_, dummy_constraint);
+
+    assert(constrained_state_space_ != nullptr);
```

Comment:
This too is extra, make_shared will throw if the allocation fails.

Reply:
done